targets/TARGET_Analog_Devices/TARGET_ADUCM302X/TARGET_ADUCM3029/bsp/rng/adi_rng.c

```diff
@@ -0,0 +1,796 @@
+            initialization (#adi_rng_Open()).  The application uses the #adi_rng_RegisterCallback()
+            API to request an application-defined callback from the RNG device driver. The RNG device
+            driver clears the interrupt when the callback exits.
+            The application callback should <b>avoid extended processing</b>
+            during callbacks as the callback is executing context of the initiating interrupt and will
+            block lower-priority interrupts.  If extended application-level interrupt processing is
+            required, the application should schedule it for the main application loop and exit the
+            callback as soon as possible.\n
+
+
+    @sa    adi_rng_Open().
+*/
+ADI_RNG_RESULT adi_rng_RegisterCallback (
+                                         ADI_RNG_HANDLE  hDevice,
+                                         ADI_CALLBACK    cbFunc,
+                                         void           *pCBParam)
+{
+    ADI_RNG_DEV_TYPE *pDevice = (ADI_RNG_DEV_TYPE*)hDevice;
+
+#ifdef ADI_DEBUG
+   if (ADI_RNG_INVALID_HANDLE(pDevice)){
+        return ADI_RNG_BAD_DEV_HANDLE;
+    }
+
+    if (NULL == pDevice->pData) {
+        return ADI_RNG_NOT_INITIALIZED;
+    }
+#endif
+
+    /* save the callback info */
+    pDevice->pData->CBFunc   = cbFunc;
+    pDevice->pData->pCBParam = pCBParam;
+
+    if (NULL != cbFunc) {
+        /* enable RNG interrupts in NVIC */
+        NVIC_EnableIRQ(pDevice->pData->IRQn);
+    } else {
+        NVIC_DisableIRQ(pDevice->pData->IRQn);
+    }
+
+    return ADI_RNG_SUCCESS;
+}
+
+/*! \cond PRIVATE */
+/* RNG driver interrupt handler. Overrides weak default handler in startup file */
+void RNG_Int_Handler(void)
+{
+    ISR_PROLOG();
+    ADI_RNG_DEV_TYPE *pDevice = &gRNG_Device[0];
+    register uint16_t candidate;
+
+    /* if we have an initialized driver... */
+    if (NULL != pDevice->pData)
+    {
+        /* if we have a registered callback */
+        if (NULL != pDevice->pData->CBFunc)
+        {
+            ADI_INT_STATUS_ALLOC();
+
+            ADI_ENTER_CRITICAL_REGION();
+            /* read status register without other interrupts in between */
+            candidate = pDevice->pRNG->STAT;
+            ADI_EXIT_CRITICAL_REGION();
+
+            /* Only have bits in stat that are necessary */
+            candidate = candidate & (BITM_RNG_STAT_STUCK | BITM_RNG_STAT_RNRDY);
+
+            while (0u != candidate) {
+                uint32_t nEvent;
+
+                if (0u != (candidate & BITM_RNG_STAT_RNRDY)) {
+                    nEvent = ADI_RNG_EVENT_READY;
+                    candidate &= (uint16_t)~BITM_RNG_STAT_RNRDY;
+                } else if (0u != (candidate & BITM_RNG_STAT_STUCK)) {
+                    nEvent = ADI_RNG_EVENT_STUCK;
+                    candidate &= (uint16_t)~BITM_RNG_STAT_STUCK;
+                } else {
+                    break;
```

Comment:
Is it robust to ignore other events silently on an unrecognised event.

Reply:
This has been discussed in the PR thread for the 4050LZ. The STUCK flag is now checked in the trng_get_bytes to make sure no hardware errors have occurred.

Reply:
This has been discussed in the PR thread for the 4050LZ. The STUCK flag is now checked in the trng_get_bytes to make sure no hardware errors have occurred.

Reply:
This has been discussed in the PR thread for the 4050LZ. The STUCK flag is now checked in the trng_get_bytes to make sure no hardware errors have occurred.